### PR TITLE
refactor: Add build GHA image

### PR DESCRIPTION
**Pull request by jesse-peters**
I took another whack at [this](https://github.com/aurae-runtime/aurae/issues/79) using a few different methods. The time to run is still a bit longer than native GHA jobs, but with a self-hosted runner I think it would be pretty similar.

I'm building the image as the first step in the pipeline, then using that image to run the jobs on downstream. I have introduced [cargo-chef](https://crates.io/crates/cargo-chef) to package the dependencies, without needing to compile the code itself.

Using minideb we launch a container that has the rust requirements (but without extras like cargo-chef). This runs the actual make steps, and uses native GHA caching for the rust modules.

I would be happy to setup and own a self-hosted runner and [docker registry](https://github.com/aurae-runtime/aurae/issues/78). I'm happy to chat about that in discord too (kuxaku) :)

Most of the build layers should cache across branches too (using buildkits native GHA cache), so the first time it runs should be the longest.

Biggest Risks:
- I got the caching wrong and the pipeline doesn't have correct dependencies someplace along the test/build process
- The build times are longer, impacting developer feedback times

Out of scope:
- I did not apply the docker image to the `deploy` step yet.

I kept the original two GHA jobs in the pipeline for reference. If/when this gets merged they can be ran alongside for a bit to ensure its working, or deleted in a second PR.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# syntax = docker/dockerfile:1.4
```
Comment (krisnova) [inline Dockerfile:1]: Do we want this to be the "root" Dockerfile for the project? Would it make sense to have a `/docker` or `/images` or `/containers` directory where we can house multiple Dockerfiles for different reasons? What about plumbing the docker commands through to the Makefile and having an `$OCI` variable so we can build with out container tools as well?

Something like this:

```make 
OCI=docker
$(OCI) build -t $(NAME) .
```

Comment (jesse-peters) [reply Dockerfile:1]: You beat me to it - I was mid writing up a comment and was going to ask if this was the desired location. I'll move it and add the docker commands via the make file too.

Comment (jesse-peters) [reply Dockerfile]: I'm not using any native docker commands in the pipeline, just pre-defined GHA jobs. I have added the ability to build the docker container to the Makefile.

**Comment by jesse-peters** [comment]
Sorry for the noise, I created this PR a little bit early it looks like

**Comment by krisnova** [comment]
Want to squash for us?

**Comment by jesse-peters** [comment]
Happy to. Bare with me, git and I dont get alone sometimes :) (I usually have squash turns on automatically in my repos)

**Comment by future-highway** [comment]
@cla-bot check

**Comment by jesse-peters** [comment]
I think I'll need a project maintainer to add write permissions containers to this project. I'm using `aurae-builder` as the name of the package. Now is a good time to change that too if we wanted to use a different one.

I think its here for an org:

https://github.com/orgs/aurae-runtime/packages

https://docs.github.com/en/packages/managing-github-packages-using-github-actions-workflows/publishing-and-installing-a-package-with-github-actions#upgrading-a-workflow-that-accesses-a-registry-using-a-personal-access-token

**Comment by jesse-peters** [comment]
@cla-bot check

**Comment by krisnova** [comment]
Hey @jesse-peters I hate to sound like the idiot who doesn't understand how to use GitHub, but I have spent the last 30 minutes trying to understand what it is I am supposed to do to help you out here and I honestly am stumped.

What exactly do you need? Can you be as descriptive as possible? Is there any way you can share a "this specific thing is failing and if you make this specific thing go away we will be happy" style description?

**Pull request closed without merging**